src/search.coffee

```diff
@@ -6,234 +6,257 @@ url = require 'url'
+          response.results.forEach (result) ->
+            unless result.id in existingIds
+              data.results.push result
+              existingIds.push result.id
+
+        callback null, data
+
+  counts: (query, options, callback) ->
+    [callback, options] = [options, {}] if _.compact(arguments).length is 2
+
+    sources = ['web', 'images', 'videos', 'news']
+    methods = {
+      web: _.bind @_rawWeb, this
+      images: _.bind @_rawImages, this
+      videos: _.bind @_rawVideos, this
+      news: _.bind @_rawNews, this
+    }
+
+    executeSearch = (source, callback) =>
+      # We only need to one result (zero is impossible) from most verticals.
+      # With a web search, however, the `totalEstimatedMatches` needs to be
+      # checked from a higher page for accurate data.
+      #
+      # The 1,000 value comes from empirical data. It seems that after 600
+      # results, the accuracy gets quite consistent and accurate. I picked 1,000
+      # just to be in the clear. It also doesn't matter if there are fewer than
+      # 1,000 results.
+      pagination = {offset: 0, top: 1}
+      if source is 'web'
+        _.each pagination, (value, key) ->
+          pagination[key] += 1000
+
+      methods[source] query, (_.defaults pagination, options), callback
+
+    async.mapLimit sources, @parallelLimit, executeSearch,
+      (err, responses) ->
+        callback err if err?
+
+        data = {}
+        data[source] = 0 for source in sources
+
+        responses.forEach (response, i) ->
+          data[sources[i]] = response.count
+
+        callback null, data
+
+  # This modifies the endpoint used for searching to retrieve params specific
+  # to separate verticals (i.e. width/height for images and runtime for videos).
+  _verticalSearch: (vertical, q, options, callback) ->
+    options = _.extend({}, options, {q, endpoint: "#{vertical}/search"})
+    @_parallelSearch options, callback
+
+  # This sends the vertical via the responseFilters query param for methods
+  # which don't have specific verticals (i.e. web, spelling, related).
+  _filteredSearch: (responseFilter, q, options, callback) ->
+    options = _.extend({}, options, {q, responseFilter})
+    @_parallelSearch options, callback
+
+  _rawWeb: (query, options, callback) ->
+    @_filteredSearch 'webpages', query, options, callback
 
   web: (query, options, callback) ->
-    @verticalSearch 'Web', _.bind(@extractWebResults, this), query, options,
-      callback
+    [callback, options] = [options, {}] if _.compact(arguments).length is 2
 
-  extractWebResults: (results) ->
-    @mapResults results, ({ID, Title, Description, DisplayUrl, Url}) ->
-      id: ID
-      title: Title
-      description: Description
-      displayUrl: DisplayUrl
-      url: Url
+    @_rawWeb query, options, (err, data) =>
+      callback err if err?
+      callback null, @_extractWebResults data
+
+  _extractWebResults: (data) ->
+    _.map data.results, (result) ->
```

Comment:
This probably doesn't make a difference here, and it's kind of a micro-optimization, but coffee's `for .. in` is perfectly readable for simple transformations like this and emits native JS code instead of whatever `_.map` uses and whatever compatibility code and indirection it has under the hood.

Also, when you're only using one or a few properties of the input object, you can do destructuring in the arguments definition. You usually want to provide a default argument as well when you destructure. Coffee will try to access properties of an undefined object if `undefined` or `null` are ever passed in.

So this could be:

```coffee
 _extractWebResults: ({results} = {}) ->
  for {id, name, snippet, url, displayUrl} in results
    {id, title: name, description: snippet, url, displayUrl}
```

If you want to go even further, you can rename variables when you destructure. So the loop comprehension could look like:

```coffee
for {id, name: title, snippet: description, url, displayUrl} in result
  {id, title, description, url, displayUrl}
```